File: 剑指offer/剑指 Offer 34. 二叉树中和为某一值的路径.cpp
//注意回溯法的顺序 不能在return时插入 因为左右两个几点会插入两次



/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    vector<vector<int> > res;
    vector<int> temp;
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        if(!root) return {};
        recursion(root, sum);
        return res;
    }
    void recursion(TreeNode *root, int sum)
    {
        if(!root) return;
        temp.push_back(root -> val);
        sum -= root -> val;
        if(sum == 0 && !root -> left && !root -> right)
            res.push_back(temp);
        recursion(root -> left, sum); // 左
        recursion(root -> right, sum); // 右
        temp.pop_back(); // 函数退出之前先弹出当前节点
    }
};

File: 剑指offer_java/剑指 Offer 34. 二叉树中和为某一值的路径.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    List<Integer> list = new ArrayList<>(); 
    public List<List<Integer>> pathSum(TreeNode root, int target) {
        if(root==null) return res;
        helper(root, target);
        return res;
    }
    void helper(TreeNode root, int target){
        if(root==null) return;
        target-=root.val;
        list.add(root.val);
        if(target== 0&& root.left==null&& root.right==null) res.add(new ArrayList(list));
        helper(root.right,target);
        helper(root.left,target);
        list.remove(list.size() - 1);
    }
}
